website/src/documentation/io/developing-io-java.md

```diff
@@ -0,0 +1,372 @@
+
+To create a data source for your pipeline, you must provide the format-specific
+logic that tells a runner how to read data from your input source, and how to
+split your data source into multiple parts so that multiple worker instances can
+read your data in parallel. If you're creating a data source that reads
+unbounded data, you must provide additional logic for managing your source's
+watermark and optional checkpointing.
+
+Supply the logic for your source by creating the following classes:
+
+  * A subclass of `BoundedSource` if you want to read a finite (batch) data set,
+    or a subclass of `UnboundedSource` if you want to read an infinite (streaming)
+    data set. These subclasses describe the data you want to read, including the
+    data's location and parameters (such as how much data to read).  
+
+  * A subclass of `Source.Reader`. Each Source must have an associated Reader that
+    captures all the state involved in reading from that `Source`. This can
+    include things like file handles, RPC connections, and other parameters that
+    depend on the specific requirements of the data format you want to read.  
+
+  * The `Reader` class hierarchy mirrors the Source hierarchy. If you're extending
+    `BoundedSource`, you'll need to provide an associated `BoundedReader`. if you're
+    extending `UnboundedSource`, you'll need to provide an associated
+    `UnboundedReader`.
+
+  * One or more user-facing wrapper composite transforms (`PTransform`) that
+    wrap read operations. [PTransform wrappers](#ptransform-wrappers) discusses
+    why you should avoid exposing your sources.
+
+
+### Implementing the Source subclass
+
+You must create a subclass of either `BoundedSource` or `UnboundedSource`,
+depending on whether your data is a finite batch or an infinite stream. In
+either case, your `Source` subclass must override the abstract methods in the
+superclass. A runner might call these methods when using your data source. For
+example, when reading from a bounded source, a runner uses these methods to
+estimate the size of your data set and to split it up for parallel reading.
+
+Your `Source` subclass should also manage basic information about your data
+source, such as the location. For example, the example `Source` implementation
+in Beam’s [DatastoreIO](https://beam.apache.org/releases/javadoc/current/index.html?org/apache/beam/sdk/io/gcp/datastore/DatastoreIO.html)
+class takes host, datasetID, and query as arguments. The connector uses these
+values to obtain data from Cloud Datastore.
+
+#### BoundedSource
+
+`BoundedSource` represents a finite data set from which a Beam runner may read,
+possibly in parallel. `BoundedSource` contains a set of abstract methods that
+the runner uses to split the data set for reading by multiple workers.
+
+To implement a `BoundedSource`, your subclass must override the following
+abstract methods:
+
+  * `splitIntoBundles`: The runner uses this method to split your finite data
```

Comment:
This method was renamed to 'split': https://github.com/apache/beam/blob/master/sdks/java/core/src/main/java/org/apache/beam/sdk/io/BoundedSource.java#L55